File: 이준영/백준/Gold/13023. ABCDE/ABCDE.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.StringTokenizer;

public class Main {
	static boolean check=false;
	public static void main(String[] args) throws Exception{
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine());
		int N = Integer.parseInt(st.nextToken());
		int M = Integer.parseInt(st.nextToken());
		boolean[] visited = new boolean[N];
		ArrayList<Integer>[] fri = new ArrayList[N];
		for(int i=0;i<N;i++) {
			fri[i]= new ArrayList<>();
		}
		for(int i=0;i<M;i++) {
			st = new StringTokenizer(br.readLine());
			int a = Integer.parseInt(st.nextToken());
			int b = Integer.parseInt(st.nextToken());
			fri[a].add(b);
			fri[b].add(a);
		}
		for(int i=0;i<N;i++) {
			if(check)
				break;
			visited[i]=true;
			btk(i,0,visited,fri);
			visited[i]=false;
		}
		if(check) {
			System.out.println(1);
		} else {
			System.out.println(0);
		}
	}
	public static void btk(int cur,int count,boolean[] visited,ArrayList<Integer>[] fri) {
		if(check)
			return;
		if(count==4) {
			check=true;
			return;
		}
		for(int i:fri[cur]) {
			if (!visited[i]) {
				visited[i]=true;
				btk(i,count+1,visited,fri);
				visited[i]=false;
			}
		}
	}
}

File: 이준영/백준/Gold/13023. ABCDE/ABCDE.py
# 각 번호에서 시작하여 이러한 관계가 생기는지를 백트래킹으로 탐방
# 방문여부는 visited 배열로 관리
def btk(cnt, cur):                      # (지금까지 길이, 현재 노드)
    global check
    if check == 1:                      # 이미 찾아서 더 찾을 필요 없음
        return

    for i in friend[cur]:
        if not visited[i]:
            if cnt == 4:                # 지금까지 4명이었는데 방문하지 않은 친구가 있다면 성공
                check = 1
                break
            else:                       # 아니면 방문하지 않은 친구 쪽으로 검사해봄
                visited[i] = True
                btk(cnt + 1, i)
                visited[i] = False


N, M = map(int, input().split())
friend = [[] for _ in range(N + 1)]
check = 0
for _ in range(M):
    a, b = map(int, input().split())
    friend[a].append(b)
    friend[b].append(a)

visited = [False] * (N + 1)
for i in range(1, N + 1):
    visited[i] = True
    btk(1, i)
    visited[i] = False

print(check)
